raiden/raiden_service.py

```diff
@@ -540,9 +541,67 @@ def connection_manager_for_token(self, token_address):
             raise InvalidAddress('token is not registered.')
         return manager
 
+    def leave_all_token_networks_async(self):
+        token_addresses = self.channelgraphs.keys()
+        leave_results = []
+        for token_address in token_addresses:
+            try:
+                connection_manager = self.connection_manager_for_token(token_address)
+            except InvalidAddress:
+                pass
+            leave_results.append(connection_manager.leave_async())
+        return leave_results
+
+    def close_and_settle(self):
+        log.info('raiden will close and settle all channels now')
+
+        connection_managers = [
+            self.connection_manager_for_token(token_address) for
+            token_address in self.channelgraphs.keys()
+        ]
+
+        def blocks_to_wait():
+            return max(
+                connection_manager.min_settle_blocks
+                for connection_manager in connection_managers
+            )
+
+        all_channels = sum(
+            [connection_manager.open_channels for connection_manager in connection_managers], []
```

Comment:
apparently the performance of this is also really terrible. will use the [`itertools` recipe](https://docs.python.org/2.7/library/itertools.html#recipes) instead. thanks!